### PR TITLE
length for chain (fixes #84)

### LINKED ISSUE #84: chain(it1,it2) should have length
... when it1 and it2 have a definite length

### PR DESCRIPTION
**Pull request by CarloLucibello**
Adds a length function for chains when they are supposed to have one

Probably this is not the right way to do if, since iteratorsize is type unstable. Does anyone have better ideas?

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -118,6 +118,16 @@ ch2 = chain(1:0, 1:2:5, 0.2:0.1:1.6)
 @test eltype(ch2) == typejoin(Int, Float64)
 @test collect(ch2) == [1:2:5; 0.2:0.1:1.6]
 
+ch3 = chain(1:10,1:10, 1:10)
+@test length(ch3) == 30
+@test Base.iteratorsize(ch3) == Base.HasLength()
```
Comment (kmsquire) [inline test/runtests.jl]: `iteratorsize` is imported (if available) and otherwise defined in this package, so it shouldn't be prefixed with `Base`. (Causing tests to fail.)

Comment (CarloLucibello) [inline test/runtests.jl]: looks like I introduced some typo at the last moment just before pushing it (Meth instead of MethodErrod), that was causing the failure

Base.iteratorsize is doing the right job on my machine, i.e. using the method defined in Iterators.jl. Should I still change the test?

Comment (CarloLucibello) [inline test/runtests.jl]: ah, I see, sorry, I was looking at 0.5 travis build

I can switch Base.Iteratorsize to Interface.Iteratorsize but we still have a problem with Base.HasLength(). Should I put the whole test inside a
```julia
if VERSION >= v"0.5.0-dev+3305"
...
end
```
?

Comment (kmsquire) [inline test/runtests.jl]: Yes, I think that sounds good.

**Comment by CarloLucibello** [comment]
the test should be good also on 0.4, thanks @kmsquire for the review. Also, I left the prefix Base., tell me if you want it to be Interators. I can't tell the difference at this point. 

Do you thinks that this should be merged even if iteratorsize becomes type unstable?
```julia
julia> c=chain(1:10,1:10)
Iterators.Chain{Tuple{UnitRange{Int64},UnitRange{Int64}}}((1:10,1:10))

julia> @code_warntype Base.iteratorsize(c)
Variables:
  #self#::Base.#iteratorsize
  x::Iterators.Chain{Tuple{UnitRange{Int64},UnitRange{Int64}}}

Body:
  begin 
      return $(Expr(:invoke, LambdaInfo for iteratorsize(::Type{Iterators.Chain{Tuple{UnitRange{Int64},UnitRange{Int64}}}}), :(Base.iteratorsize), Iterators.Chain{Tuple{UnitRange{Int64},UnitRange{Int64}}}))
  end::Union{Base.HasLength,Base.SizeUnknown}
```

**Comment by iamed2** [comment]
IMO the way forward is to determine this when constructing `Chain` and include the iterator size as a type parameter. That way `length` could also be defined only when it would return a length.

**Comment by CarloLucibello** [comment]
>IMO the way forward is to determine this when constructing Chain and include the iterator size as a type parameter. That way length could also be defined only when it would return a length.

nice! I'll try to cook something to also preserve julia 0.4 compatibility

File: test/runtests.jl [hunk 2]
```diff
@@ -124,13 +124,13 @@ if VERSION >= v"0.5.0-dev+3305"
 	@test Base.iteratorsize(ch3) == Base.HasLength()
 end
 
-r = ( x for x in 1:10 if isodd(x))
-ch4 = chain(1:10, r)
 if VERSION >= v"0.5.0-dev+3305"
+	r = ( x for x in 1:10 if isodd(x))
```
Comment (kmsquire) [inline test/runtests.jl]: Even though this wouldn't actually run on v0.4, the compiler still attempts to parse it, and unfortunately, this syntax is a parse error on v0.4. 😢 

For now, my suggestion is to comment out this test, explaining that it is a parse error on Julia v0.4, and that it should be uncommented when support for v0.4 is dropped.

**Comment by CarloLucibello** [comment]
ok, the tests are passing now, but...

on a second thought the problem of type instability has shifted from the function `iteratorsize` to the function `chain`, so I don't know if we can call that a progress

**Comment by CarloLucibello** [comment]
I figured a way out of the type instability issue, this should be good now (if tests pass :)

**Comment by CarloLucibello** [comment]
>Base.iteratorsize should be defined on the iterator type, not the instance.

will do

>_chain_is should return IsInfinite() as soon as it hits an iterator with IsInfinite().

it does (the line after #fallback)

> IteratorSize type import is unused.

will remove

> There should be more tests for more iteratorsize iterator combinations.

I'm kinda impaired here by the 0.4 compatibility requirement (e.g. cannot use generators and test for return type of itersize can be run only on 0.5)

**Comment by iamed2** [comment]
> it does (the line after #fallback)
`SizeUnknown()` and `IsInfinite()` are different. Unless maybe you didn't push a change?

> I'm kinda impaired here by the 0.4 compatibility requirement (e.g. cannot use generators and test for return type of itersize can be run only on 0.5)
Feel free to write more 0.5-only tests when necessary. There should be enough available in terms of data structures and other iterators that you shouldn't need generators.

**Comment by CarloLucibello** [comment]
>>it does (the line after #fallback)

>SizeUnknown() and IsInfinite() are different. Unless maybe you didn't push a change?

you're right, sorry, I misread

Everything should be fixed now (crossing fingers)

**Comment by oxinabox** [comment]
This is a good plan, and I like the idea of having `iteratorsize` defined on more things,
A few issues though:

It has been suggested that 
`HasLength` should only be used by things that have a `length` that can be found in `O(1)` time.
And further that `iteratorsize` itself should always be `O(1)`


This (like #60) is has the property that `iteratorsize`  and `length` are is `O(n)` when `n` is the number of iterators being chained together.

I personally think that is probably Ok, as `n` is kinda expected to be smallish.

But maybe it isn't
Consider the case of: `Chain((rand(rand(1:10)) for ii in 1:1_000)...)` for example.
Indeed, I myself often collect up sequences of words into sentences,
and then I might like to chain all the sentences into the words of the whole document.
For me, than `n` might be 100_000, or even several billion.


>IMO the way forward is to determine this when constructing Chain and include the iterator size as a type parameter

Adding that extra type parameter would let you calculate the iteratorsize at construction time -- the constructor is already `O(n)` where `n` is the number of iterators being chained.
(In 0.4 that parameter could be set to Void, I guess).
It could also let you calculate the length at construction time and store that in a field.

----------------

I note that you have define the iteratorsize recursively.
Which is very elegant (I really like defining things that way).
But julia does not support TCO, right now.
So that will be slower than a loop.
And can stackoverflow.
It should be rewritten to be done with a loop.

**Comment by CarloLucibello** [comment]
> This is a good plan, and I like the idea of having iteratorsize defined on more things,
> A few issues though:
> 
> It has been suggested that 
> HasLength should only be used by things that have a length that can be found in O(1) time.
> And further that iteratorsize itself should always be O(1)
> 
> This (like #60) is has the property that iteratorsize and length are is O(n) when n is the number of iterators being chained together.
> 
> I personally think that is probably Ok, as n is kinda expected to be smallish.
I'm not sure iteratorsize is O(n), It should be O(1) if the compiler is smart, since it's all static evaluation

Length is O(n), but is we compute it at creation time we have a problem: 
```julia
r = rand(10)
c = chain(r, r)
push!(r, 1.)
length(c) != length(collect(c)) # problem here
```

I think chain should be optimized for small n. It is a good thing also that the constructor chain(x...) enforces `n` to be small, in some sense

btw: length is also implemented for Product (still O(n))
> 
> But maybe it isn't
> Consider the case of: Chain((rand(rand(1:10)) for ii in 1:1_000)...) for example.
> Indeed, I myself often collect up sequences of words into sentences,
> and then I might like to chain all the sentences into the words of the whole document.
> For me, than n might be 100_000, or even 10_000_000
> In this case, though perhaps splatting has already killed me.
> So it may not be anything to worry about
You may want to use Base.flatten with that
> 
> IMO the way forward is to determine this when constructing Chain and include the iterator size as a type parameter
> Adding that extra type parameter would let you calculate the iteratorsize at construction time -- the constructor is already O(n) where n is the number of iterators being chained.
> (In 0.4 that parameter could be set to Void, I guess).
> It could also let you calculate the length at construction time and store that in a field.
> 
> I note that you have define the iteratorsize recursively.
> Which is very elegant (I really like defining things that way).
> But julia does not support TCO, right now.
> So that will be slower than a loop.
> And can stackoverflow.
> It should be rewritten to be done with a loop.

runtime evaluation of iteratorsize, as done with a loop, even at construction time, would cause type instability (see comments above). I fear much more that perfomancewise that recursion. Do you agree?

**Comment by oxinabox** [comment]
> I think chain should be optimized for small n. It is a good thing also that the constructor chain(x...) enforces n to be small, in some sense

I agree.
Maybe we can in fact enforce that, 
say by making chain throw up a warning if you chain more than 32 things (unless you a pass a  override kwargument?)
And that warning can direct people to some new iterator: `LongChain`,
that takes an iterator of iterators as input (see #34),
and which is `SizeUnknown`

**Comment by CarloLucibello** [comment]
I've never `@generated` anything yet :) I'll look into that

**Comment by CarloLucibello** [comment]
and here it is the third or fourth life for this pr, now with generated iteratorsize

@oxinabox do you know of any downsides of generated functions? All I could find are some, almost creepy,  "don't you use them if you don't need them"

**Comment by oxinabox** [comment]
>@oxinabox do you know of any downsides of generated functions? All I could find are some, almost creepy, "don't you use them if you don't need them"

They are spooky pseudo-macros.
Just like macro's you should not use them with you can avoid it.
But here I think they are the right tool for the job

There are a bunch of things you are not allowed to do with them.
In particular: https://github.com/JuliaLang/julia/issues/18568
but also  https://github.com/JuliaLang/julia/issues/18747

There purpose is for writing type generic code, that depends only on the types on the arguments.
It seems like they are a perfect match for traits.
(and that as you found, it is nearly impossible to write code to deal with traits  in a type stable way, without them)

**Comment by iamed2** [comment]
The downside to this is that it will compile a function every time you call `chain`. 

All these options will need some empirical measurement.

**Comment by CarloLucibello** [comment]
> The downside to this is that it will compile a function every time you call chain.

only if you chain different types. But that would happen in any case using dispatch I guess

> All these options will need some empirical measurement.

Yeah, it would be nice to set up the benchmarking framework. I'm sure this is the best approach for runtime speed. I'm not sure about compile speed, though it may be not that relevant. Merge and eventually revisit later?

File: test/runtests.jl [hunk 3]
```diff
@@ -116,6 +118,53 @@ ch2 = chain(1:0, 1:2:5, 0.2:0.1:1.6)
...
+
+ch5 = chain()
+@test length(ch5) == 0
+if VERSION >= v"0.5.0-dev+3305"
+	@test iteratorsize(ch5) == HasLength()
+end
+
+c = chain(ch1, ch2, ch3)
+@test length(c) == length(ch1) + length(ch2) + length(ch3)
+#@test collect(c) == [ch1; ch2; ch3] # why doesn't work?
```
Comment (iamed2) [inline test/runtests.jl]: Remove this commented-out line

File: test/runtests.jl [hunk 4]
```diff
@@ -106,7 +109,6 @@ end
 
 # chain
 # -----
-
```
Comment (iamed2) [inline test/runtests.jl]: This was a nice blank line.

File: test/runtests.jl [hunk 5]
```diff
@@ -1,5 +1,8 @@
+include("../src/Iterators.jl")
 using Iterators, Base.Test
-
```
Comment (iamed2) [inline test/runtests.jl]: Bring back this whitespace.

File: test/runtests.jl [hunk 6]
```diff
@@ -1,5 +1,8 @@
+include("../src/Iterators.jl")
```
Comment (iamed2) [inline test/runtests.jl]: Remove this please.

File: src/Iterators.jl [hunk 7]
```diff
@@ -105,17 +106,34 @@ done(it::RepeatCallForever, state) = false
...
 # Concatenate the output of n iterators
-
 immutable Chain{T<:Tuple}
     xss::T
 end
 
-iteratorsize{T<:Chain}(::Type{T}) = SizeUnknown()
+if VERSION >= v"0.5.0-dev+3305"
+    # corner case of empty chain
+    iteratorsize(c::Chain) = iteratorsize(typeof(c))
```
Comment (iamed2) [inline src/Iterators.jl]: Remove this; there is already a generic function in Base which does this.

File: test/runtests.jl [hunk 8]
```diff
@@ -116,6 +118,53 @@ ch2 = chain(1:0, 1:2:5, 0.2:0.1:1.6)
...
+end
+
+ch3 = chain(1:10, 1:10, 1:10)
+@test length(ch3) == 30
+if VERSION >= v"0.5.0-dev+3305"
+	@test iteratorsize(ch3) == HasLength()
+end
+
+r = countfrom(1)
+ch4 = chain(1:10, countfrom(1))
```
Comment (iamed2) [inline test/runtests.jl]: Deserves an `eltype` test.

File: test/runtests.jl [hunk 9]
```diff
@@ -116,6 +118,53 @@ ch2 = chain(1:0, 1:2:5, 0.2:0.1:1.6)
...
+c = chain(r, r)
+@test length(c) == 8
+@test collect(c) == [vec(r); vec(r)]
+if VERSION >= v"0.5.0-dev+3305"
+	@test iteratorsize(r) == HasShape()
+	@test iteratorsize(c) == HasLength()
+end
+
+if VERSION >= v"0.5.0-dev+3305"
+	r = distinct(1:10)
```
Comment (iamed2) [inline test/runtests.jl]: Use `distinct(collect(1:10))`; it's possible someone might make a no-op `UnitRange` optimization for `distinct`.

File: test/runtests.jl (hunk 8 above)
Comment (CarloLucibello) [inline test/runtests.jl]: done

Comment (CarloLucibello) [inline test/runtests.jl]: done

**Comment by CarloLucibello** [comment]
why travis is taking so long? it's been like 9 hours

**Comment by iamed2** [comment]
I'm not sure, but I think there are a finite number of `julia` workers so if a ton of Julia projects push things we'll be in a long line.

**Comment by kmsquire** [comment]
Specifically, everything in the `JuliaLang` organization is run serially. This is why a number of packages have been recently migrated over to other organizations (`JuliaMath`, JuliaIO`, etc).

**Comment by CarloLucibello** [comment]
is something missing for this?
